src/sagas/wallet.js

```diff
@@ -0,0 +1,518 @@
+      }
+
+      return [...acc, token.uid];
+    }, []);
+
+  // We don't need to wait for the metadatas response, so just fork it
+  yield fork(fetchTokensMetadata, registeredTokens);
+
+  // Since we already know here what tokens are registered, we can dispatch actions
+  // to asynchronously load the balances of each one. The `put` effect will just dispatch
+  // and continue, loading the tokens asynchronously
+  for (const token of registeredTokens) {
+    yield put(tokenFetchBalanceRequested(token));
+  }
+}
+
+/**
+ * Fetch a single token from the metadataApi
+ *
+ * @param {Array} token The token to fetch from the metadata api
+ * @param {String} network Network name
+ *
+ * @memberof Wallet
+ * @inner
+ */
+export async function fetchTokenMetadata(token, network) {
+  if (token === hathorLibConstants.HATHOR_TOKEN_CONFIG.uid) {
+    return {};
+  }
+
+  const metadataPerToken = {};
+
+  try {
+    const data = await metadataApi.getDagMetadata(
+      token,
+      network,
+    );
+
+    // When the getDagMetadata method returns null
+    // it means that we have no metadata for this token
+    if (data) {
+      const tokenMeta = data[token];
+      metadataPerToken[token] = tokenMeta;
+    }
+  } catch (e) {
+    // Error downloading metadata.
+    // TODO: We should wait a few seconds and retry if still didn't
+    // reach the retry limit
+    // eslint-disable-next-line
+    console.log('Error downloading metadata of token', token);
+  }
+
+  return metadataPerToken;
+}
+
+/**
+ * The wallet needs each token metadata to show information correctly
+ * So we fetch the tokens metadata and store on redux
+ */
+export function* fetchTokensMetadata(tokens) {
+  const { network } = yield select((state) => state.serverInfo);
+
+  const tokenChunks = chunk(tokens, METADATA_CONCURRENT_DOWNLOAD);
+
+  let tokenMetadatas = {};
+  for (const tokenChunk of tokenChunks) {
+    const responses = yield all(tokenChunk.map((token) => fetchTokenMetadata(token, network)));
+    for (const response of responses) {
+      tokenMetadatas = {
+        ...tokenMetadatas,
+        ...response,
+      };
+    }
+  }
+
+  yield put(tokenMetadataUpdated(tokenMetadatas));
+}
+
+// This will create a channel to listen for featureFlag updates
+export function* listenForFeatureFlags(featureFlags) {
+  const channel = eventChannel((emitter) => {
+    const listener = (state) => emitter(state);
+    featureFlags.on('wallet-service-enabled', (state) => {
+      emitter(state);
+    });
+
+    // Cleanup when the channel is closed
+    return () => {
+      featureFlags.removeListener('wallet-service-enabled', listener);
+    };
+  });
+
+  try {
+    while (true) {
+      const newUseWalletService = yield take(channel);
+      const oldUseWalletService = yield select((state) => state.useWalletService);
+
+      if (oldUseWalletService && oldUseWalletService !== newUseWalletService) {
+        NativeModules.HTRReloadBundleModule.restart();
+      }
+    }
+  } finally {
```

Comment:
If this forked saga gets cancelled either from the outside or the inside (using the `cancel` effect), the loop will break by throwing an cancellation error, so we can safely close the channel (and thus removing the event listener) on the `finally` block